File: app/src/main/java/io/vinter/trackmyanime/utils/TopMainRecyclerAdapter.java
package io.vinter.trackmyanime.utils;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;

import java.util.List;

import io.vinter.trackmyanime.R;
import io.vinter.trackmyanime.entity.AnimeTop;

public class TopMainRecyclerAdapter extends RecyclerView.Adapter<TopMainRecyclerAdapter.TopMainRecyclerViewHolder> {

    private Context context = null;
    private List<List<AnimeTop>>  animeTopList;
    private String[] names;
    ItemClickListener animeClickListener;

    public TopMainRecyclerAdapter(Context context, List<List<AnimeTop>> list, String[] names,
                                  ItemClickListener animeClickListener){
        this.context = context;
        this.animeTopList = list;
        this.names = names;
        this.animeClickListener = animeClickListener;
    }

    @NonNull
    @Override
    public TopMainRecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        View itemView = inflater.inflate(R.layout.item_top_main, viewGroup, false);
        return new TopMainRecyclerViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull TopMainRecyclerViewHolder holder, int i) {
        TopHorizontalRecyclerAdapter adapter = new TopHorizontalRecyclerAdapter(context,
                animeTopList.get(i).subList(0, 10), animeClickListener);
        holder.horizontalRecycler.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        holder.horizontalRecycler.setAdapter(adapter);
        holder.topTypeButton.setText(names[i]);
    }

    @Override
    public int getItemCount() {
        return animeTopList.size();
    }

    public static class TopMainRecyclerViewHolder extends RecyclerView.ViewHolder {

        RecyclerView horizontalRecycler;
        Button topTypeButton;

        public TopMainRecyclerViewHolder(@NonNull View itemView) {
            super(itemView);
            horizontalRecycler = itemView.findViewById(R.id.horizontalRecycler);
            topTypeButton = itemView.findViewById(R.id.topTypeButton);
        }
    }
}

File: app/src/main/java/io/vinter/trackmyanime/utils/TopMainRecyclerAdapter.kt
package io.vinter.trackmyanime.utils

import android.content.Context
import android.support.v4.util.Pair
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button

import java.util.Collections

import io.vinter.trackmyanime.R
import io.vinter.trackmyanime.entity.top.AnimeTop
import io.vinter.trackmyanime.utils.recycler.ChildHorizontalRecyclerView

class TopMainRecyclerAdapter(private var context: Context, private var animeTopList: List<Pair<List<AnimeTop>, String>>,
                             private var animeClickListener: (Any, Any) -> Unit) : RecyclerView.Adapter<TopMainRecyclerAdapter.TopMainRecyclerViewHolder>() {

    init {
        Collections.sort(animeTopList) { listStringPair, t1 -> listStringPair.second!!.compareTo(t1.second!!) }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): TopMainRecyclerViewHolder {
        val inflater = LayoutInflater.from(viewGroup.context)
        val itemView = inflater.inflate(R.layout.item_top_main, viewGroup, false)
        return TopMainRecyclerViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: TopMainRecyclerViewHolder, i: Int) {
        val adapter = TopHorizontalRecyclerAdapter(context,
                animeTopList[i].first!!.subList(0, 10), animeClickListener)
        val mLayoutManager = object : LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false) {
            override fun canScrollVertically(): Boolean {
                return false
            }
        }
        holder.horizontalRecycler.layoutManager = mLayoutManager
        holder.horizontalRecycler.adapter = adapter
        holder.horizontalRecycler.isNestedScrollingEnabled = false
        holder.topTypeButton.text = animeTopList[i].second
    }

    override fun getItemCount(): Int {
        return animeTopList.size
    }

    class TopMainRecyclerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        internal var horizontalRecycler: ChildHorizontalRecyclerView = itemView.findViewById(R.id.horizontalRecycler)
        internal var topTypeButton: Button = itemView.findViewById(R.id.topTypeButton)

    }
}
